zerver/lib/actions.py

```diff
@@ -2126,19 +2126,26 @@ def notify_reaction_update(
     # Update the cached message since new reaction is added.
     update_to_dict_cache([message])
 
-    # Recipients for message update events, including reactions, are
-    # everyone who got the original message.  This means reactions
-    # won't live-update in preview narrows, but it's the right
-    # performance tradeoff, since otherwise we'd need to send all
-    # reactions to public stream messages to every browser for every
-    # client in the organization, which doesn't scale.
-    #
-    # However, to ensure that reactions do live-update for any user
-    # who has actually participated in reacting to a message, we add a
-    # "historical" UserMessage row for any user who reacts to message,
-    # subscribing them to future notifications.
-    ums = UserMessage.objects.filter(message=message.id)
-    send_event(user_profile.realm, event, [um.user_profile_id for um in ums])
+    # Recipients for stream message update events, including reactions, are
+    # everyone who is subscribed to the stream the message belongs to.
+    # Credit goes to Steve Howell who helped me creating this patch.
+
+    # The user ids to send the event to.
+    user_ids: Union[Iterable[int], Iterable[Mapping[str, Any]]]
+
+    if message.recipient.type == Recipient.STREAM:
+        # Get the id of the stream the message belongs to.
+        stream_id: int = message.recipient.type_id
+        # Get all active subscriptions to this stream.
+        subscriptions: QuerySet[Subscription] = get_active_subscriptions_for_stream_id(stream_id)
+        # Extract the relevant user ids.
+        user_ids = subscriptions.filter(user_profile__is_active=True).values_list(
+            "user_profile__id", flat=True
+        )
+    else:
+        user_ids = [um.user_profile_id for um in UserMessage.objects.filter(message=message.id)]
+
+    send_event(user_profile.realm, event, user_ids)
```

Comment:
Thanks for working on this!  I think the logic here isn't quite correct, though it's close.  The problem is we're not properly accounting for situations where the stream has `!history_public_to_subscribers`, where someone might be reacting to a message sent before they joined, where our security model should not allow the new user to access those messages in any way.

* The UserMessage strategy is correct for any stream with !history_public_to_subscribers, as well as PMs.
* The `Subscription` strategy is correct for streams with history_public_to_subscribers, except that is_guest users don't get access to history from before they joined if the stream is not `is_web_public` (at least in our plans).
* So for streams with history_public_to_subscribers, we want to include all UserMessages plus any users with a Subscription that gives them access to history, which should be roughly adding `is_guest=False` to the filter.

So I think you can create a `user_ids` set that starts with `{um.user_profile_id for um in UserMessage.objects.filter(message=message.id)}` and then extend it with a somewhat complicated set of conditionals around the stream.  

Structurally the right path is probably to write a new first commits that creates a bulk-users version of `can_access_stream_history` (returning the set of user IDs with access to that stream's history), write tests for that, and then call that function from here.

Reply:
I'd call the new function `user_ids_with_stream_history_access(stream: Stream) -> Set[int]`.